File: agent_mcp/router/security_headers.py
"""Security response headers for every router response.

Threat model: open-internet exposure. The router served no security
response headers at all, so a public deploy inherited none of the
browser-side defences (clickjacking, MIME sniffing, referrer leak,
missing HSTS). This middleware adds them uniformly — including on
error responses and redirects — by wrapping the whole handler chain.

CSP notes (the one header that can break the app):

  The dashboard is a static Next.js export. Its HTML carries inline
  ``<script>`` blocks (Next's hydration runtime, ``self.__next_f``
  pushes) and the Jinja login/setup templates carry an inline
  ``<style>`` block, so a strict ``script-src 'self'`` / ``style-src
  'self'`` would break both surfaces. Static export can't use per-
  response nonces (the files are served verbatim), so we allow
  ``'unsafe-inline'`` for script + style. We deliberately do NOT add
  ``'unsafe-eval'`` — production Next.js export doesn't need it, and
  it's the more dangerous of the two.

  The parts that add real value regardless — ``frame-ancestors
  'none'`` (clickjacking), ``object-src 'none'``, ``base-uri 'self'``,
  ``default-src 'self'`` (no off-origin loads) — are all kept strict.

The whole CSP is overridable via ``AGENT_MCP_CSP`` for operators who
front the router differently or want to tighten it further.

HSTS is emitted ONLY on HTTPS requests (same trusted-proxy-gated
``X-Forwarded-Proto`` / scheme heuristic as ``login.cookie_secure_flag``)
so the plain-HTTP dev / VM smoke doesn't get pinned to HTTPS.

Cache-Control (SC-1): sensitive router surfaces — the login page, authed
API JSON, and 401/error bodies — carried no cache directive, so they
could land in bfcache or a shared cache. We stamp ``no-store`` via
``setdefault`` so those responses opt out of caching, while the static
dashboard handlers (which set their OWN ``Cache-Control`` BEFORE this
middleware runs — ``no-store`` for HTML, ``immutable`` for hash-named
assets) keep their explicit value untouched.

Server banner (SC-2 / SD-3): aiohttp fills a ``Server:
Python/… aiohttp/…`` header at prepare time (``setdefault`` on
``SERVER_SOFTWARE``). That discloses exact framework versions for
CVE-matching on a direct bind (nginx masks it in prod, but this is the
defence-in-depth gap). We assign a neutral, version-free banner
UNCONDITIONALLY (not ``setdefault``) so it's present before aiohttp's
own default fires and wins over it.

R7-F2: the middleware-level overwrite above only runs for responses
that reach ``web.Application`` — a request malformed enough to trip
aiohttp's own HTTP/1.1 parser (bad ``Content-Length``, a bad chunk
size, ...) never gets that far, so it still leaked the raw banner.
Closed at the root by reassigning aiohttp's own ``SERVER_SOFTWARE``
constant(s) at THIS module's import time — see the comment beside
that reassignment below for why two separate bindings need patching.
Residual gap accepted: the other hardened headers (CSP, X-Frame-
Options, ...) still cannot land on that parser-rejected path — there
is no ``Request``/``Application`` context at that point without
patching aiohttp's internal ``RequestHandler.handle_error``, out of
proportion for a defense-in-depth header on an already-4xx'd
malformed request. Mitigated in production by fronting the router
behind nginx/Tailscale, which doesn't forward raw malformed requests
to begin with.
"""

from __future__ import annotations

import logging
import os
from typing import Awaitable, Callable

from aiohttp import web

from .client_disconnect import client_gone_response, client_is_gone


logger = logging.getLogger(__name__)


# Pragmatic-but-protective default. See module docstring for why
# script/style get ``'unsafe-inline'`` (static Next.js export) while
# everything else stays strict.
_DEFAULT_CSP = (
    "default-src 'self'; "
    "script-src 'self' 'unsafe-inline'; "
    "style-src 'self' 'unsafe-inline'; "
    "img-src 'self' data: blob:; "
    "font-src 'self' data:; "
    "connect-src 'self'; "
    "object-src 'none'; "
    "base-uri 'self'; "
    "form-action 'self'; "
    "frame-ancestors 'none'"
)

# Minimal Permissions-Policy: the dashboard needs none of these
# powerful features, so deny them outright.
_DEFAULT_PERMISSIONS_POLICY = (
    "geolocation=(), microphone=(), camera=(), payment=(), usb=()"
)

_HSTS_VALUE = "max-age=63072000; includeSubDomains"

# Neutral, version-free replacement for aiohttp's ``Server:
# Python/… aiohttp/…`` banner (SC-2 / SD-3). Names the product but
# discloses nothing a CVE scanner can pivot on.
_SERVER_BANNER = "agent-mcp"

# R7-F2: `_apply_headers`'s unconditional `hdrs["Server"] = ...`
# above only fires for responses that flow through
# `web.Application.__call__` — i.e. this middleware. A genuinely
# malformed HTTP/1.1 request (bad/duplicate Content-Length, a bad
# chunk size, a garbage request line, ...) never gets that far:
# aiohttp's own `RequestHandler.handle_error` (web_protocol.py)
# builds and writes a bare `Response(...)` directly from within the
# protocol's `data_received`, bypassing `Application` — and every
# middleware, including this one — entirely. That response fills
# `Server` via `headers.setdefault(hdrs.SERVER, SERVER_SOFTWARE)`,
# so the only way to neutralise it there is to make aiohttp's OWN
# default neutral, by reassigning the constant it reads.
#
# `aiohttp.web_response` does `from .http import SERVER_SOFTWARE` at
# import time — that's a SEPARATE name binding in its own module
# namespace, not an attribute lookup through `aiohttp.http` — so
# patching only `aiohttp.http.SERVER_SOFTWARE` leaves
# `aiohttp.web_response.SERVER_SOFTWARE` (the one `Response.
# write_headers` actually reads) holding the real version string.
# Both must be reassigned.
#
# Doing this at THIS module's import time (rather than, say, inside
# `make_app()`) makes it a guaranteed side effect of the single
# statement both live entrypoints already have —
# `router/app.py::make_app` and `cli.py`'s `router` subcommand both
# `from .security_headers import security_headers_middleware` before
# constructing the `web.Application`, and in both cases that
# construction completes (as a function-call argument or otherwise)
# before `web.run_app` is invoked. One import, both entrypoints
# covered, no risk of a future call site forgetting a separate
# startup hook.
import aiohttp.http as _aiohttp_http
import aiohttp.web_response as _aiohttp_web_response

_aiohttp_http.SERVER_SOFTWARE = _SERVER_BANNER
_aiohttp_web_response.SERVER_SOFTWARE = _SERVER_BANNER


def _request_is_https(request: web.Request) -> bool:
    """Same heuristic as ``login.cookie_secure_flag``.

    Honours ``X-Forwarded-Proto`` first (TLS terminates upstream and
    forwards plain HTTP), but ONLY when the direct peer is a trusted
    proxy (``rate_limit.request_from_trusted_proxy``) — the header is
    client-settable, so an untrusted peer must not drive the HSTS
    decision (OBS7 class-sweep; this was the last of 5 XFH/XFP
    trust-boundary consumers to get the gate — see
    ``login.cookie_secure_flag`` / ``login._external_origin`` /
    ``mount.external_origin`` / ``sso._default_redirect_url``).
    Falls back to the direct request scheme otherwise. ``rate_limit``
    is imported lazily so this leaf middleware carries no hard import
    on it at module load time.

    Security: the exploitable direction of an ungated header here is
    HSTS-suppression — spoofing ``http`` strips HSTS from an
    otherwise-secure response, opening a downgrade window on a LATER
    connection. (Spoofing ``https`` to force HSTS onto genuine
    plain-HTTP has no client impact: RFC 6797 requires an
    already-validated TLS connection for a browser to honour the
    header at all.)
    """
    from . import rate_limit

    if rate_limit.request_from_trusted_proxy(request):
        forwarded = request.headers.get("X-Forwarded-Proto", "").lower()
        if forwarded == "https":
            return True
        if forwarded == "http":
            return False
    return request.url.scheme == "https"


def _csp() -> str:
    override = os.environ.get("AGENT_MCP_CSP")
    return override if override else _DEFAULT_CSP


def _apply_headers(response: web.StreamResponse, request: web.Request) -> None:
    """Set the security headers on ``response`` (idempotent per response).

    Uses ``setdefault`` so a handler that intentionally set its own
    value (rare) isn't clobbered.
    """
    hdrs = response.headers
    hdrs.setdefault("X-Content-Type-Options", "nosniff")
    hdrs.setdefault("X-Frame-Options", "DENY")
    hdrs.setdefault("Referrer-Policy", "strict-origin-when-cross-origin")
    hdrs.setdefault("Content-Security-Policy", _csp())
    hdrs.setdefault("Permissions-Policy", _DEFAULT_PERMISSIONS_POLICY)
    # Cross-origin isolation (defense-in-depth alongside frame-ancestors
    # / X-Frame-Options). COOP: same-origin severs cross-origin window
    # handles so a popup/opener can't reach into this context (blocks
    # a class of Spectre-style + tab-nabbing side channels). CORP:
    # same-origin refuses embedding of the router's responses as a
    # cross-origin resource. The dashboard is fully same-origin, so
    # neither breaks any legitimate load.
    hdrs.setdefault("Cross-Origin-Opener-Policy", "same-origin")
    hdrs.setdefault("Cross-Origin-Resource-Policy", "same-origin")
    # HSTS only over HTTPS — never pin a plain-HTTP dev/VM to TLS.
    if _request_is_https(request):
        hdrs.setdefault("Strict-Transport-Security", _HSTS_VALUE)
    # SC-1: opt sensitive surfaces out of caching. ``setdefault`` so the
    # static dashboard handlers keep their own explicit value (they set
    # ``Cache-Control`` on the response BEFORE this middleware runs —
    # ``no-store`` for HTML, ``immutable`` for hash-named assets).
    hdrs.setdefault("Cache-Control", "no-store")
    # SC-2 / SD-3: overwrite aiohttp's version-disclosing ``Server``
    # banner. Direct assignment (NOT ``setdefault``) because aiohttp
    # fills ``SERVER_SOFTWARE`` via its own ``setdefault`` at prepare
    # time — our value must already be present so it wins.
    hdrs["Server"] = _SERVER_BANNER


@web.middleware
async def security_headers_middleware(
    request: web.Request,
    handler: Callable[[web.Request], Awaitable[web.StreamResponse]],
) -> web.StreamResponse:
    """Attach security headers to every response, including errors.

    aiohttp raises ``HTTPException`` for redirects (login 303s) and
    error responses; those ARE the response object, so we catch, stamp
    the headers, and re-raise to cover the full surface.

    An UNhandled exception (a bare ``ValueError`` from malformed
    multipart, a ``TypeError`` in a handler, …) is NOT an
    ``HTTPException``, so it would otherwise propagate past this
    outermost middleware to aiohttp's core 500 renderer — which never
    runs ``_apply_headers`` (SD-R5-1). That path leaked aiohttp's
    version-disclosing ``Server`` banner and dropped every hardened
    header. We catch it here, log the real exception server-side, and
    return a generic 500 stamped with the full header set — keeping the
    exception detail off the wire.
    """
    try:
        response = await handler(request)
    except web.HTTPException as exc:
        _apply_headers(exc, request)
        raise
    except ConnectionError as exc:
        # The client vanished mid-request — an SSE stream whose browser
        # tab closed, an aborted fetch, an abandoned upload. aiohttp
        # signals it as ClientConnectionResetError ("Cannot write to
        # closing transport") on the write side and ConnectionResetError
        # ("Connection lost") on the read side; both subclass
        # ConnectionError.
        #
        # This is the OUTERMOST middleware, so re-raising handed the
        # exception straight to aiohttp's `RequestHandler.handle_error`,
        # which logged a full `aiohttp.server` ERROR traceback for what
        # is a peer leaving — noise indistinguishable from a real
        # failure, and the thing that made every occurrence cost a
        # triage. Absorb it instead and return the client-gone status
        # (nothing reaches a wire; the peer is gone by construction).
        #
        # Gated on the DOWNSTREAM transport actually being gone: a
        # ConnectionError raised while the client is still connected is
        # somebody else's reset (a backend/upstream fault) and keeps its
        # traceback via the generic handler below.
        if client_is_gone(request):
            logger.debug(
                "client disconnected during %s %s: %s",
                request.method, request.rel_url, exc,
            )
            response = client_gone_response()
            _apply_headers(response, request)
            return response
        logger.error(
            "Connection reset during %s %s while the client was still "
            "connected; returning generic 500",
            request.method,
            request.rel_url,
            exc_info=True,
        )
        response = web.Response(status=500, text="Internal Server Error")
        _apply_headers(response, request)
        return response
    except Exception:
        # Log the real cause server-side (with traceback) but NEVER put
        # it in the response body — a generic message keeps internals
        # (and any secret in the exception text) off the wire.
        logger.error(
            "Unhandled exception in %s %s; returning generic 500",
            request.method,
            request.rel_url,
            exc_info=True,
        )
        response = web.Response(status=500, text="Internal Server Error")
        _apply_headers(response, request)
        return response
    _apply_headers(response, request)
    return response


__all__ = ["security_headers_middleware"]

File: rust/conexus-router/src/security_headers.rs
//! Security response headers for every router response. Port of
//! `agent_mcp/router/security_headers.py` (302 LOC, Phase E2 PR 15,
//! `conexus-router-headers-misc`).
//!
//! Framework-agnostic: [`security_headers`] enumerates the intended
//! header set as plain data ([`SecurityHeader`], with an explicit
//! `overwrite` flag distinguishing the one header -- `Server` -- that
//! must win even over a handler's own value from every other header,
//! which only fills in when the handler didn't already set one). PR
//! 23's real axum middleware applies this list to an actual response
//! header map (via `Entry::or_insert` for `overwrite: false`, a plain
//! assignment for `overwrite: true`) and additionally owns the
//! exception-classification wrapper (`security_headers_middleware`'s
//! own `HTTPException`/`ConnectionError`/generic-`Exception` catch
//! ladder) -- that ladder is inherently a `tower`/axum error-handling
//! concern with no framework-agnostic equivalent to port ahead of a
//! real handler stack.
//!
//! **Re-derived, not ported**: Python's `SERVER_SOFTWARE` monkeypatch
//! (neutralising aiohttp's own version-disclosing `Server:
//! Python/... aiohttp/...` default, including the parser-rejected-
//! request path that bypasses the whole middleware chain) has no
//! `hyper`/axum equivalent to patch -- neither emits a version-
//! disclosing `Server` header by default in the first place, so the
//! vulnerability class this patch closes doesn't exist on this stack.
//! `SERVER_BANNER`/the `overwrite: true` `Server` entry are kept
//! anyway (a neutral, deliberate banner is still better than
//! whatever a fronting proxy might otherwise pass through unmodified).
#![allow(dead_code)]

/// Pragmatic-but-protective default CSP. `script-src`/`style-src`
/// allow `'unsafe-inline'` because the dashboard is a static Next.js
/// export (inline hydration `<script>`s, an inline login/setup
/// `<style>` block) that can't use per-response nonces; every other
/// directive stays strict (`frame-ancestors 'none'`, `object-src
/// 'none'`, `base-uri 'self'`, `default-src 'self'`).
pub const DEFAULT_CSP: &str = "default-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; img-src 'self' data: blob:; font-src 'self' data:; connect-src 'self'; object-src 'none'; base-uri 'self'; form-action 'self'; frame-ancestors 'none'";

/// The dashboard needs none of these powerful features; deny them
/// outright.
pub const DEFAULT_PERMISSIONS_POLICY: &str =
    "geolocation=(), microphone=(), camera=(), payment=(), usb=()";

pub const HSTS_VALUE: &str = "max-age=63072000; includeSubDomains";

/// Neutral, version-free `Server` banner.
pub const SERVER_BANNER: &str = "agent-mcp";

/// Port of `_csp`: `AGENT_MCP_CSP` overrides the default when set to
/// a non-empty value (an empty override is treated as unset, matching
/// Python's `if override else _DEFAULT_CSP` falsy check).
pub fn csp(env_override: Option<&str>) -> String {
    match env_override {
        Some(v) if !v.is_empty() => v.to_string(),
        _ => DEFAULT_CSP.to_string(),
    }
}

/// Port of `_request_is_https` (the same heuristic
/// `login::cookie_secure_flag` uses): `X-Forwarded-Proto` is honoured
/// ONLY from a trusted proxy, and only for its two recognised values
/// (`"https"`/`"http"`, case-insensitive) -- anything else (absent,
/// untrusted, or garbage) falls through to the real transport scheme.
/// The exploitable direction of an ungated header here is HSTS-
/// *suppression* (spoofing `http` strips HSTS from an otherwise-secure
/// response); spoofing `https` has no client impact since RFC 6797
/// requires an already-validated TLS connection for a browser to
/// honour the header at all.
pub fn request_is_https(
    is_trusted_proxy: bool,
    forwarded_proto: Option<&str>,
    url_scheme: &str,
) -> bool {
    if is_trusted_proxy {
        if let Some(v) = forwarded_proto {
            let lower = v.to_ascii_lowercase();
            if lower == "https" {
                return true;
            }
            if lower == "http" {
                return false;
            }
        }
    }
    url_scheme.eq_ignore_ascii_case("https")
}

/// One security header this module wants applied. `overwrite: true`
/// means "set unconditionally, even over a handler's own value" (only
/// `Server`, per Python's own `hdrs["Server"] = ...` vs. every other
/// header's `hdrs.setdefault(...)`); `overwrite: false` means "fill in
/// only if the response doesn't already carry this header".
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SecurityHeader {
    pub name: &'static str,
    pub value: String,
    pub overwrite: bool,
}

fn setdefault(name: &'static str, value: impl Into<String>) -> SecurityHeader {
    SecurityHeader {
        name,
        value: value.into(),
        overwrite: false,
    }
}

/// Port of `_apply_headers`'s full intended header set -- see this
/// module's own doc for why applying it to a real response is PR 23's
/// job.
pub fn security_headers(is_https: bool, csp_value: &str) -> Vec<SecurityHeader> {
    let mut headers = vec![
        setdefault("X-Content-Type-Options", "nosniff"),
        setdefault("X-Frame-Options", "DENY"),
        setdefault("Referrer-Policy", "strict-origin-when-cross-origin"),
        setdefault("Content-Security-Policy", csp_value),
        setdefault("Permissions-Policy", DEFAULT_PERMISSIONS_POLICY),
        // Cross-origin isolation, defense-in-depth alongside
        // frame-ancestors/X-Frame-Options: COOP severs cross-origin
        // window handles; CORP refuses cross-origin embedding of the
        // router's own responses. The dashboard is fully same-origin,
        // so neither breaks a legitimate load.
        setdefault("Cross-Origin-Opener-Policy", "same-origin"),
        setdefault("Cross-Origin-Resource-Policy", "same-origin"),
    ];
    if is_https {
        // HSTS only over HTTPS -- never pin a plain-HTTP dev/VM to TLS.
        headers.push(setdefault("Strict-Transport-Security", HSTS_VALUE));
    }
    // SC-1: opt sensitive surfaces out of caching. `setdefault`
    // semantics so the static dashboard handlers keep their own
    // explicit value (`no-store` for HTML, `immutable` for
    // hash-named assets), set BEFORE this would apply.
    headers.push(setdefault("Cache-Control", "no-store"));
    headers.push(SecurityHeader {
        name: "Server",
        value: SERVER_BANNER.to_string(),
        overwrite: true,
    });
    headers
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn csp_falls_back_to_the_default_with_no_override() {
        assert_eq!(csp(None), DEFAULT_CSP);
    }

    #[test]
    fn csp_treats_an_empty_override_as_unset() {
        assert_eq!(csp(Some("")), DEFAULT_CSP);
    }

    #[test]
    fn csp_honours_a_non_empty_override() {
        assert_eq!(csp(Some("default-src 'none'")), "default-src 'none'");
    }

    #[test]
    fn request_is_https_honours_a_trusted_forwarded_proto() {
        assert!(request_is_https(true, Some("https"), "http"));
        assert!(!request_is_https(true, Some("http"), "https"));
    }

    #[test]
    fn request_is_https_ignores_an_untrusted_forwarded_proto() {
        assert!(!request_is_https(false, Some("https"), "http"));
    }

    #[test]
    fn request_is_https_falls_through_on_a_garbage_forwarded_value() {
        assert!(!request_is_https(true, Some("carrier-pigeon"), "http"));
        assert!(request_is_https(true, Some("carrier-pigeon"), "https"));
    }

    #[test]
    fn request_is_https_falls_back_to_the_real_scheme_with_no_header() {
        assert!(request_is_https(true, None, "https"));
        assert!(!request_is_https(true, None, "http"));
    }

    #[test]
    fn security_headers_includes_hsts_only_over_https() {
        let with_https = security_headers(true, DEFAULT_CSP);
        assert!(with_https
            .iter()
            .any(|h| h.name == "Strict-Transport-Security"));
        let without_https = security_headers(false, DEFAULT_CSP);
        assert!(!without_https
            .iter()
            .any(|h| h.name == "Strict-Transport-Security"));
    }

    #[test]
    fn security_headers_marks_only_server_as_overwrite() {
        let headers = security_headers(true, DEFAULT_CSP);
        for h in &headers {
            assert_eq!(
                h.overwrite,
                h.name == "Server",
                "{} had unexpected overwrite flag",
                h.name
            );
        }
    }

    #[test]
    fn security_headers_carries_the_given_csp_value_verbatim() {
        let headers = security_headers(false, "default-src 'none'");
        let csp_header = headers
            .iter()
            .find(|h| h.name == "Content-Security-Policy")
            .unwrap();
        assert_eq!(csp_header.value, "default-src 'none'");
    }
}
